File: monitor/logger.py
"""결과 로거: 점검 결과를 JSON Lines 파일에 append + stdout 출력.

알림/자동복구는 non-goal — 로그만 남긴다.
"""

from __future__ import annotations

import json
from pathlib import Path

from .models import CheckResult, LoggingConfig


class FileResultLogger:
    """JSONL 파일 로거. 디렉터리는 없으면 생성."""

    def __init__(self, config: LoggingConfig, *, echo: bool = True) -> None:
        self._path = Path(config.file)
        self._echo = echo
        self._path.parent.mkdir(parents=True, exist_ok=True)

    def log(self, result: CheckResult) -> None:
        record = {
            "timestamp": result.timestamp,
            "target": result.target_name,
            "url": result.url,
            "action": result.action_type,
            "status": "ok" if result.ok else "fail",
            "detail": result.detail,
            "duration_ms": result.duration_ms,
        }
        line = json.dumps(record, ensure_ascii=False)
        with self._path.open("a", encoding="utf-8") as f:
            f.write(line + "\n")
        if self._echo:
            mark = "OK " if result.ok else "FAIL"
            print(
                f"[{result.timestamp}] {mark} {result.target_name} "
                f"({result.action_type}) {result.detail} "
                f"{result.duration_ms}ms"
            )

File: Alog/Sources/MonitorKit/Logger.swift
import Foundation

/// 점검 결과 로거 계약.
public protocol ResultLogger: Sendable {
    func log(_ run: CheckRun)
}

/// JSON Lines 파일 로거. 디렉터리는 없으면 생성. 알림/복구는 non-goal — 로그만.
public final class JSONLLogger: ResultLogger, @unchecked Sendable {
    private let url: URL
    private let queue = DispatchQueue(label: "monitorkit.jsonllogger")
    private let iso = ISO8601DateFormatter()

    public init(url: URL) {
        self.url = url
        try? FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
    }

    public func log(_ run: CheckRun) {
        let items: [[String: Any]] = run.items.map { item in
            [
                "action": item.actionKind.rawValue,
                "status": item.ok ? "ok" : "fail",
                "detail": item.detail,
                "duration_ms": item.durationMs,
            ]
        }
        let record: [String: Any] = [
            "timestamp": iso.string(from: run.timestamp),
            "target": run.targetName,
            "url": run.url,
            "vantage": run.vantageName,
            "status": run.ok ? "ok" : "fail",
            "duration_ms": run.durationMs,
            "items": items,
        ]
        guard
            let data = try? JSONSerialization.data(
                withJSONObject: record, options: [.sortedKeys]),
            let line = String(data: data, encoding: .utf8)
        else { return }

        queue.sync {
            let payload = Data((line + "\n").utf8)
            if let handle = try? FileHandle(forWritingTo: url) {
                defer { try? handle.close() }
                _ = try? handle.seekToEnd()
                try? handle.write(contentsOf: payload)
            } else {
                try? payload.write(to: url)
            }
        }
    }

    /// 로그 파일 경로(표시용).
    public var fileURL: URL { url }
}
